### PR TITLE
Make PKCE code generation RFC compliant

### PR DESCRIPTION
**Pull request by rectalogic**
https://tools.ietf.org/html/rfc7636

The code verifier is currently hexdigest before it is base64 encoded, this is incorrect.
pyoidc PKCE fails verification with all the providers I have tried it with (auth0, identityserver4).
Changing to not hexdigest (and also force ascii encoding per the spec, not whatever encoding the system is using) fixes PKCE.

- [X ] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [X] The documentation has been updated, if necessary.
---

### CODE REVIEW

**Comment by rectalogic** [comment]
Both client and provider changes are covered by existing tests in `tests/test_x_client.py` - test_pkce_verify_256 and test_pkce_verify_512

**Comment by tpazderka** [comment]
The fail in verification is not tested and the basic implementation (not in extensions) is not covered as well.

File: tests/test_client.py [hunk 1]
```diff
@@ -271,3 +286,42 @@ def test_valid_client_info(self):
...
+        authn_broker = AuthnBroker()
+        authn_broker.add("UNDEFINED", DummyAuthn(None, "username"))
+        _prov = Provider("as",
+                         session_db_factory('https://connect-op.heroku.com'), {},
+                         authn_broker, Implicit(), verify_client)
+
+        assert _prov.verify_code_challenge(cv, args['code_challenge'], 'S512') is True
+        resp = _prov.verify_code_challenge('XXX', args['code_challenge'])
+        assert isinstance(resp, Response)
+        assert resp.info()['status_code'] == 401
```
Comment (tpazderka) [inline tests/test_client.py]: Missing newline, that is the reason for failing quality checks.

**Comment by tpazderka** [comment]
Merged.

**Pull request closed without merging**